### PR TITLE
[WIP] Add create_orientation_from_NN

### PR DESCRIPTION
**Pull request by domist07**
Add the function create_orientation_from_NN for calculating the orientation from neighbour points of the same surface by given radius (radius-search) or fix number (knn).

Relates to issue [Calculate orientation from nearest neighbours #317](https://github.com/cgre-aachen/gempy/issues/317)

### CODE REVIEW

**Comment by Leguark** [comment]
Awesome, congratulations for you first pull request. Hopefully, I will have time later on to review it and give you some hints how to create a test for this code.

File: gempy/core/data.py [hunk 1]
```diff
@@ -6,6 +6,9 @@
 import numpy as np
 import pandas as pn
 
+from sklearn.neighbors import NearestNeighbors
```
Comment (Leguark) [inline gempy/core/data.py]: Can you move this into the function so gempy still works if people do not have installed sklearn

File: gempy/core/data.py [hunk 2]
```diff
@@ -1975,6 +1978,113 @@ def create_orientation_from_surface_points(self, surface_points: SurfacePoints,
 
         return np.array([*center, *orientation, *normal])
 
+    def create_orientation_from_NN(self, surface_points,  searchcrit):
```
Comment (Leguark) [inline gempy/core/data.py]: make NN in lower case, we are trying to be more strict with PEP

File: test/test_core/test_data_manipulation2.py [hunk 3]
```diff
@@ -89,3 +90,7 @@ def test_set_is_fault():
     assert mm.faults.faults_relations_df.loc['foo2', 'foo3'] == True
     assert mm.faults.faults_relations_df.iloc[2,3] == True
     mm.set_is_fault(['foo2'], toggle=True)
+
+
+def test_create_orientation_from_nn(interpolator):
```
Comment (Leguark) [inline test/test_core/test_data_manipulation2.py]: Fill this test, to show that the thing works!

File: gempy/core/data.py (hunk 2 above)
Comment (domist07) [reply gempy/core/data.py]: Done!

File: gempy/core/data.py (hunk 1 above)
Comment (domist07) [reply gempy/core/data.py]: Done!

**Comment by domist07** [comment]
I found a problem with faults.
If I calculate the orientations simply by Nearest Neighbours but between them is a fault you got maybe the false red vector instead the correct green vector 
Calc_Ori
To avoid this we have to split the points of each surface in groups between the faults.
I consider the following pseudo code:
```
x=1
For each surface
    For each fault
        For each point
            calculate distance to fault
            if dist >= 0
                groupID = groupID + x
        x = x*10
```
As result you got hopefully this
Calc_Ori_Map
Now you can split the points by their `groupID` and do the nearest neighbour search.

BUT: To know which surface is fault you have to define that _before_ in the gempy model. But if you create the orientations _after_ the surface definition you have to do the series and surface definition twice. Otherwise the series information is missing at the orientations. It's like a "chicken or the egg" problem. @Leguark do you have an idea how to solve the problem? Maybe you have to modify the `add_orientation` function.

**Comment by Leguark** [comment]
Hey Dominik, almost there, can you make a test as well as the notebook. If you pull and go to test/test_core/test_data_manipulation I created a function where you place your code. That runs with pytest and have quite a few advantages with respect a notebook. The input data you can place it in test/input_data. If you get stuck just push your latest changes and I will finish it :)

With respect the faults problem... it is definetly and issue, I think we can start thinking to solve it when #271 hierarchical modeling is implemented. For now I would add you function and leave it to the responsibility of the user to be careful - maybe we can document it in the notebook you just created

**Comment by domist07** [comment]
The commits for the notebook should not be visible here. It was an accident. The notebook was created for issue #322.

How will be your prefered method to test the function? With a synthetic dataset where I know the correct orientations or should I use a real dataset where I can get the normal vectors from PCA?
Have I only to check the direction of the orientations or somthing else?

**Comment by Leguark** [comment]
Usually a synthetic model works better so you can make an assert within the test. The important aspect of a test is to make sure the code works as promised and does not get broken so make it as simple as possible :)

**Comment by domist07** [comment]
In my gempy fork I have add two test functions to: [test/test_core/test_data_manipulation.py](https://github.com/domist07/gempy/blob/master/test/test_core/test_data_manipulation.py)
and I have set up a jupyter notebook here: [notebooks/Issue322/Issue318.ipynb](https://github.com/domist07/gempy/blob/master/notebooks/Issue318/Issue318.ipynb)
but I can't open and edit the notebook with jupyter nbviewer, but in GitHub it works. Could you open it?
In the notebook you have only to link the test functions. Then the test is ready.

**Comment by domist07** [comment]
The test is working! 🥳
Here is the link to the notebook:
https://nbviewer.jupyter.org/github/domist07/gempy/blob/master/notebooks/Issue318/Issue318.ipynb

**Comment by domist07** [comment]
Additionally I've made a notebook with _automatic detecting faults_ and _splitting the points in groups_ to avoid [the upper desribed problem](https://github.com/cgre-aachen/gempy/pull/318#issuecomment-606136496). It only works for flat faults!
https://github.com/domist07/gempy/blob/master/notebooks/Issue318_Faults/Ori-NN_Faults.ipynb

For that we should discuss if we integrate it in the function _create_orientation_from_nn_ or should we made a second?
The advantage of integrading it will be that we could save a few loops and checks in the function _create_orientation_from_nn_.
The advantage of a second function will be that the user has more options.
Should I made for this topic a new issue/pull request?

File: test/test_core/test_data_manipulation.py [hunk 4]
```diff
@@ -258,6 +258,27 @@ def test_rescaling(self, create_rescaling):
     def test_additional_data(self, create_additional_data):
         return create_additional_data
 
+def test_calc_ori_knn(coordinates, ref_orientations):
```
Comment (Leguark) [inline test/test_core/test_data_manipulation.py:261]: this test are not running. You need to use pytest (simply typing pytest on the console when you are on the folder)

File: notebooks/Issue322/Issue322.ipynb [hunk 5]
```diff
@@ -0,0 +1,343 @@
+{
```
Comment (Leguark) [inline notebooks/Issue322/Issue322.ipynb:1]: This notebooks they have to go away. Adding an example with the functionality would be great but it would need to be a .py file in the examples (new in GemPy 2.2)

File: notebooks/Issue322/322_Influence-surface-thickness_Points.csv [hunk 6]
```diff
@@ -0,0 +1,117 @@
+X,Y,Z,surface
```
Comment (Leguark) [inline notebooks/Issue322/322_Influence-surface-thickness_Points.csv:1]: specially for the test, it would be better if you use one of the models that are already there so we do need add more data

File: gempy/core/data.py [hunk 7]
```diff
@@ -1975,6 +1975,116 @@ def create_orientation_from_surface_points(self, surface_points: SurfacePoints,
 
         return np.array([*center, *orientation, *normal])
 
+    def create_orientation_from_nn(self, surface_points,  searchcrit):
```
Comment (Leguark) [inline gempy/core/data.py:1978]: I would split this function in one that gives you the nearest surface points to a given surface point.

It should look like something like `select_nearest_surfaces_points(surface_points, index, searchcrit) -> index`

Once we have the indices we can simple use `gp.set_orientation_from_surface_points` avoiding code repretition

File: gempy/core/data.py [hunk 8]
```diff
@@ -1975,6 +1975,116 @@ def create_orientation_from_surface_points(self, surface_points: SurfacePoints,
...
+        
+        from sklearn.neighbors import NearestNeighbors
+        from sklearn.preprocessing import normalize
+    
+        # extract surface names
+        surfaces = np.unique(surface_points['surface'])
+        nbrs = []
+        search_id = []
+        # for each surface
+        if isinstance(searchcrit, int):  # in case knn-search
```
Comment (Leguark) [inline gempy/core/data.py:2002]: Once you have the previous function that allows to create one orientation from 1 surface point you could add another function (in the same module) such as

`set_all_orientations_from_surface_points` where you loop all the surface points as you are doing down here.

Splitting into this 3 functions would help a lot to have a clear picture of what is happening and avoid repeating code too much. Also it makes easier
to test each function separetly.

**Comment by domist07** [comment]
I've rived the functions as you wished. You can see it in my fork: https://github.com/domist07/gempy
The two modyfied files are:
https://github.com/domist07/gempy/blob/master/gempy/api_modules/setters.py (functions itself, starts at line 189) and
https://github.com/domist07/gempy/blob/master/test/test_api/test_geometry_api.py (test)

**Comment by Leguark** [comment]
Hey @domist07 sorry  for missing the continuation of this pull request. This pull request is going to be difficult to fixed since it is so far away from master. Can you make a **new** pull request from the current master with the function and the test?

If you want I can do it, but it would be a shame that it is not recorded in git that you are the author
